ksqldb-engine/src/main/java/io/confluent/ksql/internal/ThroughputMetricsReporter.java

```diff
@@ -0,0 +1,260 @@
+/*
+ * Copyright 2021 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.internal;
+
+import static java.util.Objects.requireNonNull;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.KsqlException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.config.ConfigException;
+import org.apache.kafka.common.metrics.KafkaMetric;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.metrics.MetricsContext;
+import org.apache.kafka.common.metrics.MetricsReporter;
+import org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsImpl;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.kafka.common.utils.Utils.mkSet;
+
+public class ThroughputMetricsReporter implements MetricsReporter {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ThroughputMetricsReporter.class);
+    private static final String RECORDS_CONSUMED = "records-consumed";
+    private static final String BYTES_CONSUMED = "bytes-consumed";
+    private static final String RECORDS_PRODUCED = "records-produced";
+    private static final String BYTES_PRODUCED = "bytes-produced";
+    private static final Set<String> THROUGHPUT_METRIC_NAMES =
+        mkSet(RECORDS_CONSUMED, BYTES_CONSUMED, RECORDS_PRODUCED, BYTES_PRODUCED);
+    private static final Pattern NAMED_TOPOLOGY_PATTERN = Pattern.compile("(.*?)__\\d*_\\d*");
+    private static final Pattern QUERY_ID_PATTERN =
+        Pattern.compile("(?<=query_|transient_)(.*?)(?=-)");
+
+    // CHECKSTYLE_RULES.OFF: LineLength
+    //
+    private static final Map<String, Map<String, Map<MetricName, ThroughputTotalMetric>>> registeredMetrics = new HashMap<>();
+    // CHECKSTYLE_RULES.ON: LineLength
+    private static final Map<String, String> customTags = new HashMap<>();
+    private Metrics metricRegistry;
+
+    @Override
+    public void init(final List<KafkaMetric> initial) {
+    }
+
+    @Override
+    public synchronized void configure(final Map<String, ?> configMap) {
```

Comment:
I do things slightly different than in either of the two existing reporters, rather than register the metrics during configuration/initialization and then update them as they come in with `metricChange`, I just wait and only add them then (ie in `metricChange`)